src/interrupts.rs

```diff
@@ -0,0 +1,328 @@
+    pub interrupts_enabled: bool,     // Interrupts for this device enabled?
+    pub itr_rate: u32,                // Interrupt Throttling Rate
+    pub interrupt_type: u64,          // MSI or MSIX
+    pub timeout_ms: i16,              // Interrupt timeout in ms (-1 to disable timeout)
+    pub queues: Vec<InterruptsQueue>, // Interrupt settings per queue
+}
+
+pub struct InterruptsQueue {
+    pub vfio_event_fd: RawFd,           // event fd
+    pub vfio_epoll_fd: RawFd,           // epoll fd
+    pub interrupt_enabled: bool,        // Interrupt for this queue enabled?
+    pub instr_counter: u64,             // Counter to avoid unnecessary calls to elapsed time
+    pub last_time_checked: Instant,     // Last time the interrupt flag was checked
+    pub rx_pkts: u64,                   // The number of received packets since the last check
+    pub interval: u64,                  // The interval to check the interrupt flag
+    pub moving_avg: InterruptMovingAvg, // The moving average of the hybrid interrupt
+}
+
+#[derive(Default)]
+pub struct InterruptMovingAvg {
+    pub measured_rates: VecDeque<u64>, // Moving average window
+    pub sum: u64,                      // Moving average sum
+}
+
+impl Interrupts {
+    /// Setup VFIO interrupts by checking the `device_fd` for which interrupts this device supports.
+    pub fn vfio_setup_interrupt(&mut self, device_fd: RawFd) -> Result<(), Box<dyn Error>> {
+        info!("setting up VFIO interrupts");
+        for index in (0..=VFIO_PCI_MSIX_IRQ_INDEX).rev() {
+            let mut irq_info: vfio_irq_info = vfio_irq_info {
+                argsz: mem::size_of::<vfio_irq_info>() as u32,
+                index: index as u32,
+                flags: 0,
+                count: 0,
+            };
+
+            if unsafe { libc::ioctl(device_fd, VFIO_DEVICE_GET_IRQ_INFO, &mut irq_info) } == -1 {
+                return Err(format!(
+                    "failed to VFIO_DEVICE_GET_IRQ_INFO for index {}. Errno: {}",
+                    index,
+                    std::io::Error::last_os_error()
+                )
+                .into());
+            }
+
+            if (irq_info.flags & VFIO_IRQ_INFO_EVENTFD) == 0 {
+                continue;
+            }
+
+            self.interrupt_type = index;
+            return Ok(());
+        }
+
+        self.interrupt_type = 0;
+        Ok(())
+    }
+}
+
+impl InterruptsQueue {
+    /// Add the `event_fd` file descriptor to epoll.
+    pub fn vfio_epoll_ctl(&mut self, event_fd: RawFd) -> Result<(), Box<dyn Error>> {
+        let mut event: Event = Event {
+            events: libc::EPOLLIN as u32,
+            data: event_fd as u64,
+        };
+
+        let status = unsafe { libc::epoll_create1(0) };
+        if status == -1 {
+            return Err(format!(
+                "failed to epoll_create1. Errno: {}",
+                std::io::Error::last_os_error()
+            )
+            .into());
+        }
+        let epoll_fd = status as RawFd;
```

Comment:
Please use `epoll_fd` directly instead of `status`, i.e. `let epoll_fd: RawFd = unsafe { libc::epoll_create1(0) };`

Reply:
Same here => `rc` is `usize` so how can I check later in the code if `status == -1` to check for errors?

Reply:
Which `rc`?

Reply:
ahh I meant `epoll_fd` :)

Reply:
`RawFd` is `c_int` which will almost always be a `i32`. So you can check if it is `-1` :)

Reply:
ahh right my bad yes :) will change it :)